cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -187,11 +187,20 @@ static struct sc_mountinfo_entry *sc_parse_mountinfo_entry(const char *line)
 
 	char *parse_next_string_field() {
 		char *field = &entry->line_buf[0] + offset;
-		int nscanned =
-		    sscanf(line + offset, "%s %n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		offset += offset_delta;
+		if (line[offset] == ' ') {
+			// Special case for empty fields which cannot be parsed with %s.
+			*field = '\0';
+			offset += 1;
+		} else {
+			int nscanned = sscanf(line + offset, "%s%n", field,
+					      &offset_delta);
+			if (nscanned != 1)
+				return NULL;
+			offset += offset_delta;
+			if (line[offset] == ' ') {
```

Comment:
I'm unclear why is this needed again here? Can you explain?

Reply:
The test checks if the first character of the field is a space. This is only true if the field is empty and so this gets special-cased as `scanf` style functions cannot parse empty strings with `%s` format character.

Reply:
Ok, sorry, I should be more clear: my question is whether this is needed here if same test is done on line 190?

Reply:
Ah, sorry. I looked at the wrong part of the diff.

This test is to compensate the change that was done to `sscanf` call. Earlier we parsed `%s %n` and now we parse `%s%n`. This makes us scan each field correctly. For a visual look at how this behaves:

```
zyga@fyke:~/snapd/cmd$ make -j libsnap-confine-private/unit-tests && ./libsnap-confine-private/unit-tests -p /mountinfo/parse_mountinfo_entry/empty_source
gcc -DHAVE_CONFIG_H -I.    -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -D_ENABLE_FAULT_INJECTION -g -O2 -MT libsnap-confine-private/libsnap_confine_private_unit_tests-mountinfo-test.o -MD -MP -MF libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-mountinfo-test.Tpo -c -o libsnap-confine-private/libsnap_confine_private_unit_tests-mountinfo-test.o `test -f 'libsnap-confine-private/mountinfo-test.c' || echo './'`libsnap-confine-private/mountinfo-test.c
gcc -DHAVE_CONFIG_H -I.    -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -D_ENABLE_FAULT_INJECTION -g -O2 -MT libsnap-confine-private/libsnap_confine_private_unit_tests-utils-test.o -MD -MP -MF libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-utils-test.Tpo -c -o libsnap-confine-private/libsnap_confine_private_unit_tests-utils-test.o `test -f 'libsnap-confine-private/utils-test.c' || echo './'`libsnap-confine-private/utils-test.c
mv -f libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-utils-test.Tpo libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-utils-test.Po
mv -f libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-mountinfo-test.Tpo libsnap-confine-private/.deps/libsnap_confine_private_unit_tests-mountinfo-test.Po
gcc -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -D_ENABLE_FAULT_INJECTION -g -O2   -o libsnap-confine-private/unit-tests libsnap-confine-private/libsnap_confine_private_unit_tests-classic-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-cleanup-funcs-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-error-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-fault-injection-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-locking-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-mount-opt-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-mountinfo-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-privs-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-secure-getenv-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-snap-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-string-utils-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-test-utils.o libsnap-confine-private/libsnap_confine_private_unit_tests-test-utils-test.o libsnap-confine-private/libsnap_confine_private_unit_tests-unit-tests-main.o libsnap-confine-private/libsnap_confine_private_unit_tests-unit-tests.o libsnap-confine-private/libsnap_confine_private_unit_tests-utils-test.o -lglib-2.0  -Wl,-Bstatic -lcap -Wl,-Bdynamic
/mountinfo/parse_mountinfo_entry/empty_source: Input buffer (first), with offset arrow
Output buffer (second)
 ------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>################################################################################<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 --------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@#################################################################<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 --------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@#######################<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 --------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@###########<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 ----------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@-@#########<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 ----------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@###<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 -----------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@@##<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 -------------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@@rw<
>================================================================================<
Input buffer (first), with offset arrow
Output buffer (second)
 -------------------------------------------------------------------------------v
>304 301 0:45 / /snap/test-snapd-content-advanced-plug/x1 rw,relatime - tmpfs  rw<
>#############/@/snap/test-snapd-content-advanced-plug/x1@rw,relatime@@@tmpfs@@rw<
>================================================================================<
```